### PR TITLE
Refactor observable map.

### LINKED ISSUE #13: Cleanup: Deprecate parts of `ObservableMap` similar to PR12 for...
... `ObservableList`: https://github.com/dart-lang/observable/pull/12

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/dart-lang/observable/issues/13

## 0.19.0

* Refactor and deprecate `ObservableMap`-specific API
    * `ObservableMap` no longer emits `#keys` and `#values` change records
    * `ObservableMap.spy` is deprecated, becomes `.delegate` instead
* Potentially breaking: `ObservableMap` may no longer be extended

It is also considered deprecated to be notified of `length` changes.

### CODE REVIEW

File: lib/src/collections/observable_map.dart [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+///       buildBook(grades);
+///       if (names is ObservableMap<String>, int) {
+///         grades.changes.listen(updateBook);
+///       }
+///     }
+///     ```
+///
+/// *See [MapDiffer] to manually diff two lists instead*
+abstract class ObservableMap<K, V> implements Map<K, V>, Observable {
+  /// Creates a new observable map.
```
Comment (natebosch) [inline lib/src/collections/observable_map.dart:24]: What information is this adding?

File: lib/src/collections/observable_map.dart [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      result = new ObservableMap<K, V>.sorted();
+    } else {
+      result = new ObservableMap<K, V>();
+    }
+    return result;
+  }
+
+  /// Create a new observable map using [map] as a backing store.
+  factory ObservableMap.delegate(Map<K, V> map) {
+    return new _ObservableDelegatingMap<K, V>(map);
```
Comment (natebosch) [inline lib/src/collections/observable_map.dart:44]: [nit] =>

File: lib/src/collections/observable_map.dart [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  /// Creates a new observable map that contains all entries in [other].
+  ///
+  /// It will attempt to use the same backing map if the other map is either a
```
Comment (natebosch) [inline lib/src/collections/observable_map.dart]: "Same backing map type"?
As is it sounds like the same instance

File: lib/src/collections/observable_map.dart [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  @override
+  bool get hasObservers => _allChanges.hasObservers;
+
+  @override
+  void notifyChange([ChangeRecord change]) {
+    _allChanges.notifyChange(change);
+  }
+
+  @override
+  /*=T*/ notifyPropertyChange/*<T>*/(
```
Comment (natebosch) [inline lib/src/collections/observable_map.dart:96]: Should this be marked deprecated?

**Review by natebosch — APPROVED** [review]

File: lib/src/collections/observable_map.dart (hunk 1 above)
Comment (matanlurey) [reply lib/src/collections/observable_map.dart:24]: This is copied from the existing implementation.

File: lib/src/collections/observable_map.dart (hunk 2 above)
Comment (matanlurey) [reply lib/src/collections/observable_map.dart:44]: Acknowledged

File: lib/src/collections/observable_map.dart (hunk 4 above)
Comment (matanlurey) [reply lib/src/collections/observable_map.dart:96]: It should already be on `Observable` (`_ObservableDelegatingMap` is private)

File: lib/src/collections/observable_map.dart (hunk 3 above)
Comment (matanlurey) [reply lib/src/collections/observable_map.dart]: Done, good catch.

**Review by jmesserly — APPROVED** [review]
LGTM

**Pull request merged**